File: app/src/main/java/com/example/dailyupdate/repositories/GitHubRepository.java
package com.example.dailyupdate.repositories;

import androidx.lifecycle.LiveData;

import com.example.dailyupdate.data.model.GitHubRepo;
import com.example.dailyupdate.networking.GitHubApiClient;

import java.util.List;

public class GitHubRepository {

    private static GitHubRepository instance;
    private GitHubApiClient apiClient;

    public static GitHubRepository getInstance() {
        if (instance == null) {
            instance = new GitHubRepository();
        }
        return instance;
    }

    private GitHubRepository() {
        apiClient = GitHubApiClient.getInstance();
    }

    /**
     * Get the current LiveData object
     **/
    public LiveData<List<GitHubRepo>> getGitHubRepoList() {
        return apiClient.getGitHubRepositories();
    }

    /**
     * Launch the Api query
     **/
    public void searchRepositoriesFromGitHubApi(String searchKeyword, String sortBy) {
        apiClient.queryGitHubApi(searchKeyword, sortBy);
    }

    /**
     * Launch the Api query with an additional "sortOrder" criteria
     **/
    public void searchRepositoriesFromGitHubApiWithOrder(String searchKeyword, String sortBy,
                                                         String sortOrder) {
        apiClient.queryGitHubApiWithOrder(searchKeyword, sortBy, sortOrder);
    }

}

File: app/src/main/java/com/example/dailyupdate/repositories/GitHubRepository.kt
package com.example.dailyupdate.repositories

import androidx.lifecycle.LiveData
import com.example.dailyupdate.data.models.GitHubRepo
import com.example.dailyupdate.data.models.GitHubResponse
import com.example.dailyupdate.networking.NetworkBoundResource
import com.example.dailyupdate.networking.RetrofitInstance
import com.example.dailyupdate.utilities.Resource
import kotlinx.coroutines.CoroutineScope

class GitHubRepository {

    fun searchRepositoriesFromGitHubApiWithOrder(searchKeyword: String,
                                                 sortBy: String,
                                                 sortOrder: String,
                                                 scope: CoroutineScope)
            : LiveData<Resource<List<GitHubRepo>>> {
        return object : NetworkBoundResource<List<GitHubRepo>, GitHubResponse>(scope) {
            override fun processResponse(response: GitHubResponse): List<GitHubRepo> {
                return response.gitHubRepo
            }

            override suspend fun createCall(): GitHubResponse {
                return RetrofitInstance.getGitHubService().getGitHubRepoList(searchKeyword,
                        sortBy, sortOrder)
            }

            override fun customErrorMsg(): String? {
                return "Could not retrieve GitHub repositories"
            }

        }.asLiveData()
    }

    companion object {
        val instance = GitHubRepository()
    }
}